### PR TITLE
Sort entries by primary reading

### PR DESCRIPTION
**Pull request by khaitruong922**
- Add `primary_reading` param.
- When click on gloss link, set furigana as `primary_reading` if exists.
- Entries with reading matches `primary_reading` will be sorted to top.

Works correctly for Jitendex gloss link format. Don't know if there are any kind of dictionary format which will break this feature, since the reading is depended on the classes naming.

### CODE REVIEW

**Comment by khaitruong922** [comment]
Screenshot 2024-10-17 at 17 50 52

Screenshot 2024-10-17 at 17 51 22

**Comment by khaitruong922** [comment]
Ready for review

- Extract furigana from gloss link and add `prioritized_reading` search param

image
image


- Sort entries which match `prioritized_reading` to top

image

image

**Comment by stephenmk** [comment]
Thank you for working on this. It looks good.

> Works correctly for Jitendex gloss link format. Don't know if there are any kind of dictionary format which will break this feature, since the reading is depended on the classes naming.

I don't like this approach. I think it would be better to rely on the dictionary data to specify the `?primary_reading=` parameter. I don't think it's a good idea to try to parse the primary reading from the furigana. This increases the code complexity and could lead to unexpected behavior. It's possible that someone may want to manually specify a `?primary_reading=` parameter that is not equivalent to the furigana readings within the link.

In other words, I would remove this block of code:

https://github.com/yomidevs/yomitan/blob/051a2f0eab7e6113ce83178bdf215d61219f4aa2/ext/js/display/structured-content-generator.js#L459-L481

**Comment by Kuuuube** [comment]
> I think it would be better to rely on the dictionary data to specify the ?primary_reading= parameter.

Agreed.

**Comment by khaitruong922** [comment]
The only place that user could specify the `primary_reading` is the search page, by modifying the url. I don't think this is the main use case, and no user will be aware of it.

The main purpose of this PR is to extract the reading from furigana, then sort the terms with that reading on top. For example, suppose that my dictionary has 縁 (えん) above 縁 (ふち). When I click on a word link 縁 (ふち), the entry 縁 (ふち) should be above 縁 (えん).

As far as I know, only Jitendex has furigana for word link. I agree that the implementation is quite fragile, we can find better ways to handle this.

**Comment by stephenmk** [comment]
I just published a [new version of Jitendex](https://github.com/stephenmk/stephenmk.github.io/releases/latest/download/jitendex-yomitan.zip) that includes the `primary_reading` parameter in hyperlinks.

**Comment by khaitruong922** [comment]
Tested with new version of Jitendex. Works great!

https://github.com/user-attachments/assets/96cd1c1e-f0b8-43a8-91b8-50b81e89fa64

File: test/utilities/translator.js [hunk 1]
```diff
@@ -136,6 +137,7 @@ export function createFindTermsOptions(dictionaryName, optionsPresets, optionsAr
         sortFrequencyDictionary: typeof sortFrequencyDictionary !== 'undefined' ? sortFrequencyDictionary : null,
         sortFrequencyDictionaryOrder: typeof sortFrequencyDictionaryOrder !== 'undefined' ? sortFrequencyDictionaryOrder : 'ascending',
         removeNonJapaneseCharacters: typeof removeNonJapaneseCharacters !== 'undefined' ? removeNonJapaneseCharacters : false,
+        primaryReading: typeof primaryReading !== 'undefined' ? primaryReading : '',
```
Comment (MarvNC) [inline test/utilities/translator.js:140]: would this work the same? wondering why the ones above it  explicitly check the type
```suggestion
        primaryReading: primaryReading ?? '',
```

**Review by MarvNC — APPROVED** [review]
so much work needed to just pass variables around
StampNyaa_12858561_雪子

File: test/utilities/translator.js (hunk 1 above)
Comment (khaitruong922) [reply test/utilities/translator.js:140]: `??` also checks for null, should work the same but I don't think it is needed since `primaryReading` is `string | undefined`

Comment (MarvNC) [reply test/utilities/translator.js:140]: actually wouldn't this mean that it can be returned as null if it was null to start with? when we'd probably want it to be a string in 100% of cases.

Comment (MarvNC) [reply test/utilities/translator.js:140]: not a big deal but i really wonder why nutbread decided to do it this way for the ones above it.

Comment (khaitruong922) [reply test/utilities/translator.js:140]: This is the typing for test inputs. Either the test `option` does not specify the `primaryReading` or has `primaryReading` as a string. It cannot be null unless we explicitly put null for `primaryReading` here (which is not a valid test case).

```
        {
            "name": "Find terms using primary reading 1",
            "func": "findTerms",
            "mode": "split",
            "text": "自重",
            "options": [
                "default",
                {
                    "type": "terms",
                    "primaryReading": "じちょう"
                }
            ]
        }
 ```

Comment (khaitruong922) [reply test/utilities/translator.js:140]: > not a big deal but i really wonder why nutbread decided to do it this way for the ones above it.

some props have explicit null value so probably that's the reason
```
        sortFrequencyDictionary: typeof sortFrequencyDictionary !== 'undefined' ? sortFrequencyDictionary : null,
```
